src/sqlfluff/core/parser/grammar/base.py

```diff
@@ -515,13 +515,53 @@ class BracketInfo(NamedTuple):
-                            seg_buff = match.unmatched_segments
-                            continue
+                            # Found an end bracket. Does its type match that of
+                            # the innermost start bracket (e.g. ")" matches "(",
+                            # "]" matches "[".
+                            start_index = start_brackets.index(
+                                type(bracket_stack[-1].bracket)
+                            )
+                            end_index = end_brackets.index(matcher)
+                            bracket_types_match = start_index == end_index
+                            if bracket_types_match:
+                                # Yes, the types match. So we've found a
+                                # matching end bracket. Pop the stack and carry
+                                # on.
+                                bracket_stack.pop()
+                                pre_seg_buff += pre
+                                pre_seg_buff += match.matched_segments
+                                seg_buff = match.unmatched_segments
+                                continue
+                            else:
+                                # The types don't match. Check whether the end
+                                # bracket is a definite bracket.
+                                end_is_definite = end_definite[end_index]
+                                if not end_is_definite:
+                                    # The end bracket whose type didn't match
+                                    # the innermost open bracket is not
+                                    # definite. Assume it's not a bracket and
+                                    # carry on.
+                                    pre_seg_buff += pre
+                                    pre_seg_buff += match.matched_segments
+                                    seg_buff = match.unmatched_segments
+                                else:
+                                    # Definite end bracket does not match the
+                                    # innermost start bracket. Was the innermost
+                                    # start bracket definite? If yes, error. If
+                                    # no, assume it was not a bracket.
+                                    # Can we remove any brackets from the stack which aren't definites
+                                    # to resolve the issue?
+                                    for idx in range(len(bracket_stack) - 1, -1, -1):
+                                        if not bracket_stack[idx].is_definite:
+                                            del bracket_stack[idx]
+                                            # We don't change the string buffer, we assume that was ok.
+                                            break
+                                    else:
+                                        raise SQLParseError(  # pragma: no cover
```

Comment:
This line is not covered by tests. I don't know how to create a test that hits it, but I'd feel uncomfortable removing this code because it will help with error reporting if we **do** encounter this situation someday. Any thoughts or concerns?

Reply:
Let me try to add something to hit this, I guess we should have some test to verify that we do throw an exception when brackets aren't balanced

Reply:
https://github.com/barrywhart/sqlfluff/pull/5 maybe that would do it?